tests/project/test_task1.py

```diff
@@ -19,5 +23,144 @@ def test_P1():
     pass
 
 
+def check_production(G: Graph, e: Node):
+    Is = list(filter(lambda n: n.level == 1 and n.label == "I", G.neighbors(e)))
+
+    assert len(Is) == 1
+    I = Is[0]
+
+    Es = list(filter(lambda n: n.level == 1, G.neighbors(I)))
+    assert len(Es) == 4
+
+    check_square(G, Es, I)
+
+
+def check_square(G: Graph, Es: List[Node], I: Node):
+    for E in Es:
+        assert E in G.neighbors(I)
+
+    level_0 = list(filter(lambda n: n.level == 0 and n.label == "e", G.nodes))
+    assert len(level_0) == 1
+
+    for E in Es:
+        for neighbour in G.neighbors(E):
+            if neighbour.label == "E":
+                assert neighbour in Es
+                assert bool(neighbour.x == E.x) != bool(neighbour.y == E.y)
+            else:
+                assert neighbour.label == "I"
+                E_not_neighbours = [E1 for E1 in Es if E1 not in G.neighbors(E) and E1 != E]
+                assert len(E_not_neighbours) == 1
+                E_not_neighbour = E_not_neighbours[0]
+                assert (E_not_neighbour.x + E.x) / 2 == neighbour.x
+                assert (E_not_neighbour.y + E.y) / 2 == neighbour.y
```

Comment:
You can simply use `is_node_between` utility function like so:
```suggestion
                assert is_node_between(E_not_neighbour, neighbour, E)
```

Reply:
eehhh..